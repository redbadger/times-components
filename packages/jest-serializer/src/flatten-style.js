import { StyleSheet } from "react-native-web";

export default (accum, node) => {
  const flattened = StyleSheet.flatten(node.props.style);

  const style = Object.keys(flattened || {}).length ? { style: flattened } : {};

  return {
    accum,
    props: {
      ...node.props,
      ...style
    }
  };
};
