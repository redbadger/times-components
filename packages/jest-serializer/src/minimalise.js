import { StyleSheet } from "react-native-web";
import omit from "lodash.omit";
import omitBy from "lodash.omitby";

function getType({ type }) {
  if (type instanceof Function) {
    return type.name.toLowerCase();
  }
  if (typeof type === "string") {
    return type.toLowerCase();
  }
  return typeof type;
}

function cleanSvgProps(node, props) {
  const type = getType(node);
  return type === "path" || type === "polygon" || type === "svg"
    ? omit(props, ["d", "path", "points", "viewBox"])
    : props;
}

export default excludeProps => (accum, node) => {
  const { style: styles, ...other } = node.props;

  const flattened = StyleSheet.flatten(styles);
  const style = Object.keys(flattened || {}).length ? { style: flattened } : {};

  return {
    accum,
    props: omitBy(
      {
        ...cleanSvgProps(node, other),
        ...style
      },
      excludeProps
    )
  };
};
