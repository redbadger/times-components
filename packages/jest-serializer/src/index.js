import minimalise from "./minimalise";
import { transformProps as rnwTransform, printer as rnwPrinter } from "./rnw";
import flattenStyleTransform from "./flatten-style";
import traverse from "./traverse";

const print = (serialize, accum, element) => serialize(element);

const flattenStyle = traverse(flattenStyleTransform, print);

const minimalWebTransform = minimalise(
  value => value === undefined || typeof value === "function"
);

const minimalWeb = traverse(minimalWebTransform, print);

const minimalNativeTransform = minimalise(
  (value, key) =>
    value === undefined || typeof value === "function" || key === "className"
);

const minimalNative = traverse(minimalNativeTransform, print);

const rnw = includeStyleProps =>
  traverse(rnwTransform(includeStyleProps), rnwPrinter);

const compose = (printer, ...transformers) => {
  return traverse(
    (accum, node) =>
      transformers.reduce(
        ({ accum: a, props: p }, transformer) =>
          transformer(a, { ...node, props: p }),
        {
          accum,
          props: node.props
        }
      ),
    printer
  );
};

const minimalRnw = includeStyleProps =>
  compose(rnwPrinter, minimalWebTransform, rnwTransform(includeStyleProps));

export {
  compose,
  flattenStyle,
  flattenStyleTransform,
  minimalNative,
  minimalNativeTransform,
  minimalRnw,
  minimalWeb,
  minimalWebTransform,
  print,
  rnw,
  rnwPrinter,
  rnwTransform
};
