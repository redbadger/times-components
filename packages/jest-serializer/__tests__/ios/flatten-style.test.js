import React from "react";
import { StyleSheet, Text } from "react-native";
import renderer from "react-test-renderer";
import { flattenStyle } from "../../src";

describe("The flatten style serializer should", () => {
  describe("for native", () => {
    expect.addSnapshotSerializer(flattenStyle);

    it("not remove className", () => {
      const styles = StyleSheet.create({
        colored: {
          color: "red"
        },
        padded: {
          padding: 1
        }
      });

      const component = <Text style={[styles.colored, styles.padded]} />;
      const tree = renderer.create(component);

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
