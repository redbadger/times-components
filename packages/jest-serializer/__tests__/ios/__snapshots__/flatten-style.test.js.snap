// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The flatten style serializer should for native not remove className 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Object {
      "color": "red",
      "padding": 1,
    }
  }
/>
`;
